Simplify invoices form

JevgenijVisockij opened the pull request:
<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Simplify invoices form, change the way errors are processed for options form. Intial work was done when simplifying administration form, this continues on that and brings some improvements.
| Type?         | refacto
| Category?     | BO
| BC breaks?    | yes 
| Deprecations? | no
| Fixed ticket? | Part of #16482
| How to test?  |  This PR simplifies the invoices options form. See below

## 👓  How to test 

Everything must look the same with following exceptions:
- Blue help boxes were replaced with help text under input
- Some fields now appear as required as they always were.

Also added new errors for all text fields in invoice settings page. Now they should throw errors if you put any HTML tags into text fields.

Please test that values saved are exactly the same values you put in text fields.
(Before it would not throw error about HTML tags, simple removed them from the text, so that's something I wanted to fix). 

This PR affects Administration page, how it throws errors. All text fields in administration page allow only numbers so it should be tested they they throw errors correctly for text.

## :notebook: **BC Break**
Backwards compatibility break introduced due to extension of TranslationAwareType instead of using translator as dependency. This means if any module extends GeneratePdfByDateType or GenerateByDateType files they will get an exception.

## Delayed

Since this PR includes some breaking changes we need to wait for the future 9.0 release to finish it, so it is temporarily delayed.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/20243)
<!-- Reviewable:end -->

JevgenijVisockij committed:
Added translation for help's

JevgenijVisockij committed:
Removed some accidentally deleted divs

JevgenijVisockij committed:
Added a ne theme for generate_by_status

JevgenijVisockij commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Invoices/Themes/generate_by_status_theme.html.twig:
```diff
@@ -0,0 +1,46 @@
+ *#}
+
+{% extends 'PrestaShopBundle:Admin/TwigTemplateForm:prestashop_ui_kit.html.twig' %}
+
```
Tried to figure out another way other then creating separate file. However doing the way it was done before doesn't work because then we need to extend prestashop_ui_kit and then it will start trying to display all the blocks from the theme which fails.

One way that works is creating separate theme file, couldn't find a way to do it without that.

matks replied:
Looks like you found a solution since then :)

JevgenijVisockij committed:
Added BC break notices

JevgenijVisockij committed:
Changed from additional theme to two themes

JevgenijVisockij committed:
Fixed the wrong theme

JevgenijVisockij commented:
Can't be tested before https://github.com/PrestaShop/PrestaShop/pull/20204/ is merged.

JevgenijVisockij committed:
Removed required for some fields. Added &nbsp; to dispaly space that was there before

JevgenijVisockij commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Invoices/Blocks/generate_by_status.html.twig:
```diff
@@ -64,11 +57,12 @@
         {% set badgeType = orders_count > 0 ? 'success' : 'danger' %}
         <span class="badge badge-{{ badgeType }}">{{ orders_count }}</span>
+        &nbsp;
         <span class="status-name">
```
After simplification whitespace no longer appeared between badge and status-name.

matks replied:
👍

JevgenijVisockij commented:
The tests have not passed but on further inspection only test on PHP 7.2 for Travis failed. And the error is 
"Cannot find module 'playwright' ".

So probably just need to rerun Travis for tests to pass.

LouiseBonnard commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php:
```diff
@@ -79,35 +79,74 @@ public function __construct(
+            ->add('enable_product_images', SwitchType::class, [
+                'required' => false,
+                'label' => $this->trans('Enable product image', 'Admin.Orderscustomers.Feature'),
+                'help' => $this->trans('Adds an image in front of the product name on the invoice.', 'Admin.Orderscustomers.Help'),
```
@Junebyun, some help texts are written in the third person singular while others are not. It might be nice to have it harmonized. :-)

LouiseBonnard replied:
```suggestion
                'help' => $this->trans('Add an image in front of the product name on the invoice.', 'Admin.Orderscustomers.Help'),
```

LouiseBonnard commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php:
```diff
@@ -79,35 +79,74 @@ public function __construct(
+            ->add('use_disk_cache', SwitchType::class, [
+                'required' => false,
+                'label' => $this->trans('Use the disk as cache for PDF invoices', 'Admin.Orderscustomers.Feature'),
+                'help' => $this->trans('Saves memory but slows down the PDF generation.', 'Admin.Orderscustomers.Help'),
```
```suggestion
                'help' => $this->trans('Note that it saves memory but slows down the PDF generation.', 'Admin.Orderscustomers.Help'),
```

matks commented:
Code is good but Travis is not happy. I restart Travis

Julievrz commented:
@JevgenijVisockij Were you able to check Louise's wording suggestion? :)

JevgenijVisockij commented:
> @JevgenijVisockij Were you able to check Louise's wording suggestion? :)

Yeah commited them just now, they make sense.

But this PR is on hold for a moment, I need to sort out issues with translatable type when simplifying form. What would be best course of action to put it on hold? Just put WIP in front of it?

Julievrz commented:
> > @JevgenijVisockij Were you able to check Louise's wording suggestion? :)
> 
> Yeah commited them just now, they make sense.
> 
> But this PR is on hold for a moment, I need to sort out issues with translatable type when simplifying form. What would be best course of action to put it on hold? Just put WIP in front of it?

Thank you @JevgenijVisockij ! I added the WIP label for you

JevgenijVisockij commented:
> > > @JevgenijVisockij Were you able to check Louise's wording suggestion? :)
> > 
> > 
> > Yeah commited them just now, they make sense.
> > But this PR is on hold for a moment, I need to sort out issues with translatable type when simplifying form. What would be best course of action to put it on hold? Just put WIP in front of it?
> 
> Thank you @JevgenijVisockij ! I added the WIP label for you

Thanks!

matks commented:
> But this PR is on hold for a moment, I need to sort out issues with translatable type when simplifying form. What would be best course of action to put it on hold? Just put WIP in front of it?

@JevgenijVisockij I believe we're not blocked anymore ? 😉

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Removed uneeded postProcess

JevgenijVisockij commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php:
```diff
@@ -79,35 +81,103 @@ public function __construct(
             ])
             ->add('invoice_number', NumberType::class, [
                 'required' => false,
+                'constraints' => [
```
I think we could use this opportunity to add better wording here. Currently it checks if invoice number is larger then 0 and if it's larger then last invoice number. Maybe error message should reflect that instead of plain "Invalid invoice number".

matks replied:
@Julievrz @LouiseBonnard what do you think of this idea?

Today there is a little logic that says "Invalid invoice number"
- if submitted invoice number is below zero (obviously wrong)
- if submitted invoice number is too high, greater than last invoice number

Do you think you could give us 2 wordings, one for each case, so it's more accurate for merchant?

Julievrz replied:
Hi! @JevgenijVisockij and @matks I agree that the wording should be more specific. However, I am not sure to understand the 2nd point. Indeed, I have tried to write "3000" for the invoice number. When I download the invoice the number is 3000. Then, I tried to set "4000" and there was no error message even though 4000 is greater than 3000 (the last invoice number) 
And when I download the second invoice the number is 4000 (instead of 3001 and an error message)

JevgenijVisockij replied:
Slight correction.
Error is thrown if
1. Invoice number is below zero(correct).
2. Invoice number is not greater then last invoice number. So in your example @Julievrz  now that you changed invoice number to 4000 and created the invoice you shouldn't be be able to set this setting to 4000 or anything less then that.

Julievrz replied:
I suggest "Invoice number is invalid. Please enter a positive integer greater than the last invoice number."

Julievrz replied:
Or even better, "Invoice number is invalid. Please enter a positive integer greater than the last invoice number (#%number%)." in Admin.Orderscustomers.Notification.

Wdyt? @JevgenijVisockij

JevgenijVisockij replied:
Yeah, makes sense to me!

JevgenijVisockij commented:
> > But this PR is on hold for a moment, I need to sort out issues with translatable type when simplifying form. What would be best course of action to put it on hold? Just put WIP in front of it?
> 
> @JevgenijVisockij I believe we're not blocked anymore ? wink

Yeah this is not. 

Made quite a few changes to PR but it's ready for review now.

My main 2 concerns are: 
1) Currently I am validating all text fields whether they have HTML or PHP tags(which would be removed by strip_tags should they be saved). I would love an oppinion on how I implemented this validation(via new constrain NoTags) and if there is a better way to do it. Also this probably needs a better error message.
2) I moved validation from OrderInvoiceDataProvider to Constraints in formType, should I keep the validation in OrderInvoiceDataProvider just in case or are constraints enough?

matks commented on src/Core/ConstraintValidator/NoTagsValidator.php:
```diff
@@ -0,0 +1,51 @@
+ */
+class NoTagsValidator extends ConstraintValidator
+{
+    public function validate($value, Constraint $constraint)
```
This can be (and must) be unit tested 😉

matks replied:
See https://github.com/symfony/validator/blob/5.x/Tests/Constraints/EmailValidatorTest.php on how to test ConstraintValidators (there is a dedicated TestCase for this)

JevgenijVisockij replied:
Test makes total sense! Added it.

matks commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsDataProvider.php:
```diff
@@ -40,21 +40,13 @@ final class InvoiceOptionsDataProvider implements FormDataProviderInterface
      */
     public function __construct(
-        DataConfigurationInterface $invoiceOptionsConfiguration,
-        $nextInvoiceNumber
```
Can we keep this data, but not use it? It will be a useless constructor param but this way it's not a BC break. Avoiding a BC break is good if the cost is little 😊

JevgenijVisockij replied:
Alright, makes sense.

matks commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsDataProvider.php:
```diff
@@ -70,33 +62,6 @@ public function getData()
-     *
-     * @return array Array of errors if any
-     */
-    private function validate(array $data)
```
I dont know if it's good to remove this validation. 🤔  indeed it's a duplicate validation with what the Form does. But these are 2 classes, a module developer could choose to replace the FormType by another implementation and use this one.

I'm fifty/fifty on this.

@PierreRambaud @jolelievre wdyt ? we have moved this validation to FormType using Constraints, shall we keep it nonetheless at this level or remove it?

JevgenijVisockij replied:
A good point that InvoiceOptionsDataProvider could be used not just with formType. 

Problem with validation in dataProvider is that it's harder to affect it by the module. Hook is no longer enough and you need to decorate the service, but not having validation here means data provider can't be reliably used outside the FormType. 

Also now that you mentioned it, it makes sense DataProvider should be responsible for validating what it saves not the formType. Right now I would be for having validation both in constraints if possible and in DataProvider, even if it makes harder to remove certain validation for module developers.

matks reviewed (commented):
Little feedback ; overall great job!

JevgenijVisockij committed:
Returned back the invoice number to maintain comptability

JevgenijVisockij committed:
Modified comment to reflect changed parameters

Progi1984 commented on src/Core/ConstraintValidator/Constraints/NoTags.php:
```diff
@@ -0,0 +1,46 @@
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\ConstraintValidator\Constraints;
```
```suggestion
 */
declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\ConstraintValidator\Constraints;
```

Progi1984 commented on src/Core/ConstraintValidator/NoTagsValidator.php:
```diff
@@ -0,0 +1,61 @@
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\ConstraintValidator;
```
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\ConstraintValidator;
```

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php:
```diff
@@ -108,17 +114,39 @@ public function generatePdfByStatusAction(Request $request)
+     *
+     * @return Response|null
+     */
+    protected function renderForm($byDateForm, $byStatusForm, $optionsForm, $request)
```
```suggestion
    protected function renderForm(FormInterface $byDateForm, FormInterface $byStatusForm, FormInterface $optionsForm, Request $request): ?Response
```

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php:
```diff
@@ -127,21 +155,24 @@ public function processAction(Request $request)
      * @param Request $request
      *
-     * @return bool false if an error occurred, true otherwise
+     * @return FormInterface false
```
```suggestion
     * @return FormInterface
```

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php:
```diff
@@ -127,21 +155,24 @@ public function processAction(Request $request)
+     *
+     * @throws \Exception
      */
     private function processForm(FormHandlerInterface $formHandler, Request $request)
```
```suggestion
    private function processForm(FormHandlerInterface $formHandler, Request $request): FormInterface
```

Progi1984 commented on src/PrestaShopBundle/Resources/config/services/bundle/form/form_data_provider.yml:
```diff
@@ -25,7 +25,7 @@ services:
         arguments:
             - '@prestashop.adapter.order.invoice_options.configuration'
-            - '@=service("prestashop.adapter.data_provider.order_invoice").getNextInvoiceNumber()'
+            - '@?'
```
?

Progi1984 commented on tests/Unit/Core/ConstraintValidator/NoTagsValidatorTest.php:
```diff
@@ -0,0 +1,75 @@
+        ;
+    }
+
+    protected function createValidator()
```
```suggestion
    protected function createValidator(): NoTagsValidator
```

Progi1984 reviewed (commented):
Some feedback

JevgenijVisockij commented:
> Some feedback

Fixed those, thanks!

khouloudbelguith commented:
Hi @JevgenijVisockij,

I have a bug following these steps:

0. Create an invoice from an old order
1. Go to BO > Orders > Invoices page
2. Try to Disable Invoice
3. Save
4. See an error

![image](https://user-images.githubusercontent.com/16067358/102111788-1e746500-3e37-11eb-8154-807ecb16eeae.png)

Without PR => OK
https://drive.google.com/file/d/1x03nWR7zyR-V05LPnzOAxFIKTmkWMZ-N/view?usp=sharing
Thanks!

PierreRambaud commented on tests/Unit/Core/ConstraintValidator/NoTagsValidatorTest.php:
```diff
@@ -0,0 +1,75 @@
+        ;
+    }
+
+    protected function createValidator(): NoTagsValidator
```
Some tests are missing.

What about invalid string, number (because of is_string check), invalid HTML string. And what about using an invalid constraint?

JevgenijVisockij commented:
Hey @Julievrz 

There is an error message here and in some other cases "The field is invalid. Please enter a positive integer number.". It's displayed when you enter not an integer or negative number. 
But in this case error message isn't accurate because 0 is a valid value for the field, but 0 is not a positive number. 

Do you think something like "The field is invalid. Please enter a positive integer number or 0" or The field is invalid. Please enter integer number greater or equal to 0" would make sense?

Julievrz commented:
> Hey @Julievrz
> 
> There is an error message here and in some other cases "The field is invalid. Please enter a positive integer number.". It's displayed when you enter not an integer or negative number.
> But in this case error message isn't accurate because 0 is a valid value for the field, but 0 is not a positive number.
> 
> Do you think something like "The field is invalid. Please enter a positive integer number or 0" or The field is invalid. Please enter integer number greater or equal to 0" would make sense?

"The field is invalid. Please enter an integer number greater or equal to 0." => I think this solution would be good.

Julievrz commented:
> > Hey @Julievrz
> > There is an error message here and in some other cases "The field is invalid. Please enter a positive integer number.". It's displayed when you enter not an integer or negative number.
> > But in this case error message isn't accurate because 0 is a valid value for the field, but 0 is not a positive number.
> > Do you think something like "The field is invalid. Please enter a positive integer number or 0" or The field is invalid. Please enter integer number greater or equal to 0" would make sense?
> 
> "The field is invalid. Please enter an integer number greater or equal to 0." => I think this solution would be good.

However, what about telling the user that the number should be greater than the last invoice number?

JevgenijVisockij commented:
> > > Hey @Julievrz
> > > There is an error message here and in some other cases "The field is invalid. Please enter a positive integer number.". It's displayed when you enter not an integer or negative number.
> > > But in this case error message isn't accurate because 0 is a valid value for the field, but 0 is not a positive number.
> > > Do you think something like "The field is invalid. Please enter a positive integer number or 0" or The field is invalid. Please enter integer number greater or equal to 0" would make sense?
> > 
> > 
> > "The field is invalid. Please enter an integer number greater or equal to 0." => I think this solution would be good.
> 
> However, what about telling the user that the number should be greater than the last invoice number?

Riiight. 

So there is an interesting case. Invoice number can also be 0 which I didn't consider before so now we have 2 levels of validation.

1. Constraints. Here Symfony doesn't allow you to use conditional validation, so I can't check if Invoice number is greater than last invoice number **OR** 0. So I left the validation here to make sure the user at very least enters an integer that's at least 0 or greater. The error message "The field is invalid. Please enter an integer number greater or equal to 0." is for this.

2. Data Provider Validation which happens only if constraint passes. I put back that validation that was there before my changes(so it checks correctly for number to be 0 or greater then last invoice number). I guess it would make sense to update error message here tough. 
The last one we decided on was  "Invoice number is invalid. Please enter a positive integer greater than the last invoice number." I don't know if I should use it. 
On one hand, it's not exactly correct because you can still enter 0 and the error message doesn't say that.
On the other hand, it's noted in help right below the field that you can enter 0 to and use latest one so maybe the user won't be confused by this error message.

Julievrz commented:
> > > > Hey @Julievrz
> > > > There is an error message here and in some other cases "The field is invalid. Please enter a positive integer number.". It's displayed when you enter not an integer or negative number.
> > > > But in this case error message isn't accurate because 0 is a valid value for the field, but 0 is not a positive number.
> > > > Do you think something like "The field is invalid. Please enter a positive integer number or 0" or The field is invalid. Please enter integer number greater or equal to 0" would make sense?
> > > 
> > > 
> > > "The field is invalid. Please enter an integer number greater or equal to 0." => I think this solution would be good.
> > 
> > 
> > However, what about telling the user that the number should be greater than the last invoice number?
> 
> Riiight.
> 
> So there is an interesting case. Invoice number can also be 0 which I didn't consider before so now we have 2 levels of validation.
> 
> 1. Constraints. Here Symfony doesn't allow you to use conditional validation, so I can't check if Invoice number is greater than last invoice number **OR** 0. So I left the validation here to make sure the user at very least enters an integer that's at least 0 or greater. The error message "The field is invalid. Please enter an integer number greater or equal to 0." is for this.
> 2. Data Provider Validation which happens only if constraint passes. I put back that validation that was there before my changes(so it checks correctly for number to be 0 or greater then last invoice number). I guess it would make sense to update error message here tough.
>    The last one we decided on was  "Invoice number is invalid. Please enter a positive integer greater than the last invoice number." I don't know if I should use it.
>    On one hand, it's not exactly correct because you can still enter 0 and the error message doesn't say that.
>    On the other hand, it's noted in help right below the field that you can enter 0 to use latest invoice number so maybe the user won't be confused by this error message.

@JevgenijVisockij Here is my suggestion: 

Modify the help text a bit to make it clearer: "The next invoice number will begin with this number, and then increase with each additional invoice. Enter a positive integer greater than the last invoice number or set to 0 if you want to keep the current number (which is #number)" in Admin.OrdersCustomers.Help

First error message: "Invoice number must be a positive integer, or 0 if you want to keep the current number." in Admin.OrdersCustomers.Notification

Second error message: "Invoice number must be greater than the last invoice number, or 0 if you want to keep the current number." in Admin.OrdersCustomers.Notification

JevgenijVisockij commented:
> Invoice number must be greater than the last invoice number, or 0 if you want to keep the current number.

Hey, makes sense to me. Implemented those changes.

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php:
```diff
@@ -79,35 +81,97 @@ public function __construct(
+                            'message' => $this->trans(
+                                'Invoice number must be a positive integer, or 0 if you want to keep the current number.',
+                                'Admin.Orderscustomers.Notification',
+                                ['%number%' => $this->nextInvoiceNumber]
```
```suggestion
```

Progi1984 reviewed (commented):
A small feedback

khouloudbelguith commented:
Hi @JevgenijVisockij,

This PR is ok :heavy_check_mark: 
But there are some tests NOK.
This test is `TEST_PATH="functional/BO/02_orders/02_invoices/03_invoiceOptions/05_enableDisableCurrentYear.js" URL_FO=http://prestashop/PrestaShop/ npm run specific-test`
If this PR is merged, nightly will be broken
[withPR.pdf](https://github.com/PrestaShop/PrestaShop/files/5972480/withPR.pdf)
**Without PR**=> ok
[WithoutPR.pdf](https://github.com/PrestaShop/PrestaShop/files/5972498/WithoutPR.pdf)

Ping @PrestaShop/qa-automation what do you think?

Thanks!

nesrineabdmouleh committed:
Fix UI tests

nesrineabdmouleh commented:
> Ping @PrestaShop/qa-automation what do you think?

@khouloudbelguith thanks for running the UI tests, I just fixed the problem :heavy_check_mark: 
 Please recheck!
 
Thanks

PierreRambaud commented on src/Core/ConstraintValidator/NoTagsValidator.php:
```diff
@@ -0,0 +1,63 @@
+            throw new UnexpectedTypeException($constraint, NoTags::class);
+        }
+
+        if (!$value) {
```
So, you accept false, null, and array value? :thinking:

JevgenijVisockij replied:
Good point. Changed to just checking if it's empty string.

JevgenijVisockij committed:
Added form errors for generating invoice pdf

JevgenijVisockij committed:
Added proper errors for invoice by status generation

JevgenijVisockij committed:
Changed errors in administration form to be strings instead of int

JevgenijVisockij committed:
Removed uneeded space

JevgenijVisockij commented on src/Core/ConstraintValidator/Constraints/NoTags.php:
```diff
@@ -0,0 +1,48 @@
+
+    /**
+     * {@inheritdoc}
+     */
```
I don't really need to use this since constraints don't work in this form. Do you feel this constraint is something useful for PrestaShop and I should keep it or should I remove it?

JevgenijVisockij committed:
CS Fixer fixes

JevgenijVisockij commented on src/PrestaShopBundle/Form/Exception/InvalidConfigurationDataError.php:
```diff
@@ -52,17 +52,17 @@ class InvalidConfigurationDataError
      * @param string $fieldName
      * @param int|null $languageId
      */
-    public function __construct(int $errorCode, string $fieldName, ?int $languageId = null)
```
I have changes errorCode to string i.e. `error_not_numeric_or_lower_than_zero` to make error codes more unique. With this PR I realized that having error codes as 1,2, 3 could be very problematic when you try to return error by error code but you have multiple dataProviders which then must have different error codes.

Progi1984 reviewed (changes requested):
Could you bring us a more precise description ?

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php:
```diff
@@ -121,6 +132,138 @@ public function processAction(Request $request)
+            case InvoicesByDateDataProvider::ERROR_INVALID_DATE_TO:
+            case InvoicesByDateDataProvider::ERROR_INVALID_DATE_FROM:
+            return $this->trans(
+                    'Invalid "%s" date.',
```
I don't understand what the variable will be replaced with? Could you provide an example, please? Thanks in advance!

JevgenijVisockij replied:
So before we had ```Invalid "From" date.``` and ```Invalid "To" date.``` error messages, this is basically the same thing but variable is label of the field. 
![image](https://user-images.githubusercontent.com/47781757/111492542-6dd37580-8745-11eb-9227-a49f01511361.png)
I figured it's the same message so I would join them into one. But it's not a problem to have two messages (like it was before) without variable if that's preferable.

Julievrz replied:
That's perfectly fine! I wasn't sure to understand at first but that's way clearer now, thanks!

Julievrz commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php:
```diff
@@ -121,6 +132,138 @@ public function processAction(Request $request)
+                    $lang = $langRepository->findOneBy(['id' => $error->getLanguageId()]);
+
+                    return $this->trans(
+                        'Field "%s" in language "%s" is invalid. Field must not contain HTML tags.',
```
```suggestion
                        'The "%s" field in "%s" is invalid. HTML tags are not allowed.',
```

What do you think of my suggestion? I feel like we would better say 'The Footer text field in English is invalid.' rather than 'Field Footer text in language English is invalid."

Julievrz replied:
Just want to make sure it only has to do with 'Footer text' and 'Legal free text'. If yes 'The "%s"' is working.

JevgenijVisockij replied:
Makes perfect sense, thanks! Added it.

JevgenijVisockij committed:
Change must contain html tags error message

JevgenijVisockij commented:
> Could you bring us a more precise description ?

Hey, updated it!

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php:
```diff
@@ -121,6 +135,8 @@ public function processAction(Request $request)
         return $this->redirectToRoute('admin_order_invoices');
     }
 
+
+
```
```suggestion
```

JevgenijVisockij committed:
CS fixer fixes

Progi1984 reviewed (changes requested):
:pray: Could you add some unit tests for : 
* `src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php`

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Changed Form to FormInterface

JevgenijVisockij committed:
Change the message a bit

JevgenijVisockij committed:
Changed the error message

JevgenijVisockij committed:
Changed the domain of error message

Julievrz commented on src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php:
```diff
@@ -0,0 +1,201 @@
+                );
+            case InvalidConfigurationDataError::ERROR_NOT_NUMERIC_OR_LOWER_THAN_ZERO:
+                return $this->translator->trans(
+                    '%s is invalid. Please enter an integer greater or equal to 0.',
```
```suggestion
                    '%s is invalid. Please enter an integer greater than or equal to 0.',
```

Julievrz commented on src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php:
```diff
@@ -0,0 +1,201 @@
+                    $lang = $this->langRepository->findOneBy(['id' => $error->getLanguageId()]);
+
+                    return $this->translator->trans(
+                        'The "%s" field in "%s" is invalid. Field must not contain HTML tags.',
```
```suggestion
                        'The "%s" field in %s is invalid. HTML tags are not allowed',
```

Following your comment in another PR, I don't think we need quotation marks for the language here. :)

Julievrz commented on src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php:
```diff
@@ -0,0 +1,201 @@
+                    );
+                }
+                return $this->translator->trans(
+                    'The field "%s" is invalid. HTML tags are not allowed.',
```
```suggestion
                    'The "%s" field is invalid. HTML tags are not allowed.',
```

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php

Adding than to error wording

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php

Removing quotation marks from language and changing message a little

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php

Changing message to fit with rest

Co-authored-by: Julie Varisellaz <70583503+Julievrz@users.noreply.github.com>

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Added unit test for ErrorMessageFactory

JevgenijVisockij committed:
CS fixer fixes

Progi1984 commented on src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php:
```diff
@@ -0,0 +1,201 @@
+        );
+    }
+
+    protected function getLabel(FormInterface $form, string $fieldName)
```
```suggestion
    protected function getLabel(FormInterface $form, string $fieldName): string
```

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php

Adding returnType

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Fixing phpstan errors

JevgenijVisockij committed:
Cs fixer fix

JevgenijVisockij committed:
Added back bool returns

JevgenijVisockij committed:
Cs fixer fix

khouloudbelguith commented:
Hi @JevgenijVisockij,

>Also please test that values saved are exactly the same values you put in text fields.

I have an issue with the `Invoice number` filed
1. Go to BO > Orders > Invoices page
2. Set `Invoice number` with `<<<` or characters `test`
3. Save
4. A green alert is displayed => NOK, we should have an alert to inform that "HTML tags are not allowed."

Also, there is no check when we have long characters with or without PR.

Thank you!

JevgenijVisockij committed:
Removed uneeded dump

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsDataProvider.php:
```diff
@@ -82,21 +84,63 @@ public function setData(array $data)
+     * @throws TypeException|DataProviderException
      */
-    private function validate(array $data)
+    private function validate(array $data): void
```
Could you add unit test on this class, please ?

JevgenijVisockij replied:
Added the tests

Progi1984 reviewed (commented):
A last request and it's ok

JevgenijVisockij committed:
Added tests for invoiceDataProvider

JevgenijVisockij committed:
CS fixer fix

khouloudbelguith commented:
Hi @JevgenijVisockij,

This PR needs to be reviewed by our devs.

Thank you!

Progi1984 commented on src/Core/ConstraintValidator/Constraints/NoTags.php:
```diff
@@ -1,48 +0,0 @@
-/**
- * Validation constraint for making sure there are no HTML or PHP tags to be removed by strip_tags
- */
-class NoTags extends Constraint
```
BC Break

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -198,6 +192,10 @@ protected function getNotificationsFormHandler(): FormHandlerInterface
     /**
+     * @return array
+     *
+     * @throws FieldNotFoundException
+     *
```
```suggestion
```

Progi1984 reviewed (commented):
CI is :red_circle:. Please check errors.

JevgenijVisockij committed:
Adding no tags constraint back

Progi1984 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -162,7 +155,8 @@ protected function processForm(Request $request, FormHandlerInterface $formHandl
                 $formHandler->save($data);
             } catch (DataProviderException $e) {
-                $this->flashErrors($this->getErrorMessages($e->getInvalidConfigurationDataErrors()));
+                $errorMessageFactory = $this->get('form.invalid_configuration_error_message_factory');
```
@matks Is it new behavior ?
Error Messages are not anymore in the controller ?

matks replied:
What's important I think is not to have the error messages inside the Controller class, but to do the mapping between "catched Exception" and "error message" at the controller level. We want exceptions thrown in the logic below the controller to bubble up to the controller so that the controller can decide what it wants to do with it
- let the Exception un-catched
- catch the Exception and display an error message instead
- catch the Exception and silent it (user does not see it)

We want it to happen in the Controller so that module developers that would like to modify this behavior can do it by overriding/decorating part of the controller.

So IMO change is good because
- the try/catch is in the Controller
- the choice to display message `$this->flashErrors()` is in the Controller

Where we store the error messages... this is not dictated by our architecture 😊

Progi1984 reviewed (commented):
A question for @matks

PierreRambaud commented on src/PrestaShopBundle/Form/InvalidConfigurationErrorMessageFactory.php:
```diff
@@ -0,0 +1,200 @@
+
+declare(strict_types=1);
+
+namespace PrestaShopBundle\Form;
```
Shouldn't be something like that 
```suggestion
namespace PrestaShopBundle\Form\ErrorMessage;
```

And class name `InvalidConfigurationFactory`

I don't understand why it's directly in the Form directory :thinking:

JevgenijVisockij replied:
Wasn't exactly sure where to put it and it belonged to form in general. But yeah more specific namespace probably makes sense.

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij commented:
I believe this one is ready for reviews

atomiix commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Invoices/Blocks/generate_by_date.html.twig:
```diff
@@ -24,7 +24,7 @@
 
 {% trans_default_domain "Admin.Orderscustomers.Feature" %}
-{% import '@PrestaShop/Admin/macros.html.twig' as ps %}
+{% form_theme generateByDateForm 'PrestaShopBundle:Admin/TwigTemplateForm:prestashop_ui_kit.html.twig' %}
```
```suggestion
{% form_theme generateByDateForm '@PrestaShop/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig' %}
```

atomiix commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Invoices/Blocks/generate_by_status.html.twig:
```diff
@@ -24,9 +24,7 @@
-{% import '@PrestaShop/Admin/macros.html.twig' as ps %}
-
-{% form_theme generateByStatusForm _self %}
+{% form_theme generateByStatusForm 'PrestaShopBundle:Admin/TwigTemplateForm:prestashop_ui_kit.html.twig' _self %}
```
```suggestion
{% form_theme generateByStatusForm '@PrestaShop/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig' _self %}
```

atomiix commented on src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Invoices/Blocks/invoice_options.html.twig:
```diff
@@ -24,7 +24,7 @@
 
 {% trans_default_domain "Admin.Orderscustomers.Feature" %}
-{% import '@PrestaShop/Admin/macros.html.twig' as ps %}
+{% form_theme invoiceOptionsForm 'PrestaShopBundle:Admin/TwigTemplateForm:prestashop_ui_kit.html.twig' %}
```
```suggestion
{% form_theme invoiceOptionsForm '@PrestaShop/Admin/TwigTemplateForm/prestashop_ui_kit.html.twig' %}
```

atomiix reviewed (changes requested):
Just 3 comments about the way twig files are required (that won't work anymore when #25052 will be merged), but good PR overall!

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php:
```diff
@@ -162,7 +155,8 @@ protected function processForm(Request $request, FormHandlerInterface $formHandl
             } catch (DataProviderException $e) {
-                $this->flashErrors($this->getErrorMessages($e->getInvalidConfigurationDataErrors()));
+                $errorMessageFactory = $this->get('form.error_message.invalid_configuration_error_message_factory');
+                $this->flashErrors($errorMessageFactory->getErrorMessages($e->getInvalidConfigurationDataErrors(), $form));
```
I think this part could be moved into a protected method in `FrameworkBundleAdminController` so it can be used in other controllers

jolelievre commented on tests/Unit/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsDataProviderTest.php:
```diff
@@ -0,0 +1,132 @@
+                2 => '<html>',
+            ],
+        ];
+        $this->expectException(DataProviderException::class);
```
I think you should not only expect the exception but also check the content to be sure the error collection is correctly filled You can catch the error and assert its type and content to do that

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php:
```diff
@@ -79,35 +91,72 @@ public function __construct(
-            ->add('enable_tax_breakdown', SwitchType::class)
-            ->add('enable_product_images', SwitchType::class)
-            ->add('invoice_prefix', TranslatableType::class, [
+            ->add(static::FIELD_ENABLE_INVOICES, SwitchType::class, [
```
I don't see the interest in storing these fields into constants here We never do that in any form Besides the constants are not used somewhere else so they lose all their interest

JevgenijVisockij replied:
I think I wanted to use constants when getting label but found another way. Will remove this.

Edit: Actually I use some of those consts in InvoiceOptionsDataProvider. I just replaced all of the names with consts for the sake of consistency.

jolelievre reviewed (commented):
First review, just a few comments

One thing though that worries is how the InvalidConfigurationDataError works, I like the principle but what worries me with the current architecture is that:
- the constant will only increase with time
- they will be stored in a single place
- how would a module add its own messages since everything is hardcoded

There fore the `InvalidConfigurationErrorMessageFactory` will only get bigger and bigger, it lacks flexibility, separation of domains and customizability.

I don't have the solution right now, and weekend is coming ^^ But I'll think about it on Monday ;)

jolelievre commented:
Hi @JevgenijVisockij so I thought a little about a more flexible architecture and here's my proposal:

- create a new `ConfigurationErrorFactoryInterface` (I shorten the name a little)
- `InvalidConfigurationErrorMessageFactory` (which should maybe be renamed into `ConfigurationErrorFactory` for short would act as an aggregator of other services that implement the interface
- each configuration form (or maybe each domain/page) has its own service that implements the interface, like for example `InvoicesConfigurationErrorFactory`
- each domain also has its dedicated error class that stores the CONST, the all extend `InvalidConfigurationDataError`, this will avoid overloading a single class and allow to keep things a bit cleaner and more separated
- thus for each domain we will have to add a couple of Error class and Factory class
- finally, to simplify integration and to allow modules to integrate in this, the `ConfigurationErrorFactory` which is the aggregator of all interfaces will need a list of interfaces, it will be built by relying on the tag system like what was done for command builders (see example below)

```yaml
    form.error_message.configuration_error_factory:
        class: PrestaShopBundle\Form\ErrorMessage\ConfigurationErrorFactory
        arguments:
            - !tagged core.configuration_error_factory

    form.error_message.configuration_error_factory:
        class: PrestaShopBundle\Form\ErrorMessage\InvoicesConfigurationErrorFactory
        arguments:
            - '@translator'
            - '@prestashop.core.admin.lang.repository'
        tags: ['core.configuration_error_factory']
```

That would be the idea, let me know what you think about it and if you want us to discuss and improve it (naming is just a suggestion for now). Also @matks what do you think about this idea?

PierreRambaud commented:
Hi @JevgenijVisockij ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

PierreRambaud commented:
Oups, wrong manipulation :sweat_smile:

JevgenijVisockij committed:
Adjusted unit test

JevgenijVisockij committed:
CS Fixer fixes

Progi1984 commented on src/Core/Form/ErrorMessage/ConfigurationErrorInterface.php:
```diff
@@ -0,0 +1,50 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
```
```suggestion
```

Progi1984 commented on src/Core/Form/ErrorMessage/Factory/ConfigurationErrorFactoryInterface.php:
```diff
@@ -0,0 +1,42 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
```
```suggestion
```

Progi1984 commented on src/PrestaShopBundle/Form/ErrorMessage/LabelProvider.php:
```diff
@@ -0,0 +1,68 @@
+    {
+        $view = $form->createView();
+        foreach ($view->children as $child) {
+            if ($fieldName === $child->vars['name']) {
```
```suggestion
            if ($fieldName === $child->vars['name']) {
```

Please check `'name'` key :)

JevgenijVisockij replied:
Could you elaborate? I don't quite get the problem here.

matks replied:
My guess is that Progi1984 is worried if `$child->vars` does not contain `name` key.

But maybe such children are object generated by Symfony and the `name` key inside ` $child->vars` array is provided by Symfony by default?

Progi1984 replied:
```suggestion
            if (isset($child->vars['name']) && $fieldName === $child->vars['name']) {
```

Progi1984 commented on src/PrestaShopBundle/Form/ErrorMessage/Factory/ConfigurationErrorFactory.php:
```diff
@@ -0,0 +1,108 @@
+        $messages = [];
+
+        foreach ($errors as $error) {
+            $label = $this->labelFactory->getLabel($form, $error->getFieldName());
```
```suggestion
            $label = $this->labelFactory->getLabel($form, $error->getFieldName());
            $errorMessage = null:
```

Progi1984 commented on src/PrestaShopBundle/Form/ErrorMessage/Factory/InvoiceConfigurationErrorFactory.php:
```diff
@@ -0,0 +1,124 @@
+                    [],
+                    'Admin.Orderscustomers.Notification'
+                );
+            case InvoicesConfigurationError::ERROR_CONTAINS_HTML_TAGS:
+                if ($error->getLanguageId()) {
```
Could you add unit test for checking this behavior ?

Progi1984 commented on src/PrestaShopBundle/Form/ErrorMessage/Factory/ConfigurationErrorFactory.php:
```diff
@@ -0,0 +1,108 @@
+     *
+     * @throws FieldNotFoundException
+     */
+    public function getErrorMessages(ConfigurationErrorCollection $errors, FormInterface $form): array
```
Could you add unit test for checking this behavior ?

Progi1984 commented on src/PrestaShopBundle/Form/ErrorMessage/Factory/CommonConfigurationErrorFactory.php:
```diff
@@ -0,0 +1,99 @@
+    public function getErrorMessageForConfigurationError(ConfigurationErrorInterface $error, string $label): ?string
+    {
+        switch ($error->getErrorCode()) {
+            case InvoicesConfigurationError::ERROR_CONTAINS_HTML_TAGS:
+                if ($error->getLanguageId()) {
```
Could you add unit test for checking this behavior ?

Progi1984 reviewed (commented):
@JevgenijVisockij Thank you for your contribution. Could you add some tests for checking some behaviro ?

JevgenijVisockij committed:
Fixed unit tests for error message factories

JevgenijVisockij committed:
Adding label provider test

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij committed:
Setting errorMessage as null in case it's not found

JevgenijVisockij committed:
Added separate tests for Common configuratino error factory and invoice configuration error factory

JevgenijVisockij commented:
> @JevgenijVisockij Thank you for your contribution. Could you add some tests for checking some behaviro ?

Thanks for review! I was still adding tests, everything should be covered now.

Progi1984 commented on tests/Unit/PrestaShopBundle/Form/ErrorMessage/LabelProviderTest.php:
```diff
@@ -0,0 +1,61 @@
+        $labelProvider = new LabelProvider();
+        self::assertSame('Field', $labelProvider->getLabel($form, 'field'));
+        self::assertSame('Some field name', $labelProvider->getLabel($form, 'other_field'));
+    }
```
Could you test `Exception` ?

JevgenijVisockij replied:
Done

JevgenijVisockij committed:
Adding test for field not found exception

JevgenijVisockij committed:
CS fixer fixes

Progi1984 commented on tests/Unit/PrestaShopBundle/Form/ErrorMessage/Factory/InvoiceConfigurationErrorFactoryTest.php:
```diff
@@ -0,0 +1,95 @@
+
+class InvoiceConfigurationErrorFactoryTest extends TestCase
+{
+    public function testGetErrorMessageForConfigurationError()
```
```suggestion
    public function testGetErrorMessageForConfigurationError(): void
```

Progi1984 replied:
Please check `null` value.

Progi1984 commented on tests/Unit/PrestaShopBundle/Form/ErrorMessage/Factory/CommonConfigurationErrorFactoryTest.php:
```diff
@@ -0,0 +1,97 @@
+
+class CommonConfigurationErrorFactoryTest extends TestCase
+{
+    public function testGetErrorMessageForConfigurationError()
```
```suggestion
    public function testGetErrorMessageForConfigurationError():void
```

Progi1984 commented on tests/Unit/PrestaShopBundle/Form/ErrorMessage/Factory/CommonConfigurationErrorFactoryTest.php:
```diff
@@ -0,0 +1,97 @@
+            'The "field" field is invalid. HTML tags are not allowed.',
+            $result
+        );
+    }
```
Could you check null return too, please ?

JevgenijVisockij committed:
Adding assertions for null

JevgenijVisockij committed:
Apply suggestions from code review

Adding void return types

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Fixed test not working

matks commented on src/Core/Form/ErrorMessage/Factory/ConfigurationErrorFactoryInterface.php:
```diff
@@ -0,0 +1,40 @@
+     *
+     * @return string
+     */
+    public function getErrorMessageForConfigurationError(ConfigurationErrorInterface $error, string $label): ?string;
```
What a beautiful fully typed signature ! 👩‍🎨

matks commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/GenerateByDateType.php:
```diff
@@ -26,25 +26,46 @@
  * Class generates "By Date" form
  * in "Sell > Orders > Invoices" page.
+ *
+ * Backwards compatibility break introduced in 1.7.8.0 due to extension of TranslationAwareType
```
```suggestion
 * Backwards compatibility break introduced in 8.0.0 due to extension of TranslationAwareType
```
Now 😅  it's 8.0.0.

matks commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/GenerateByStatusType.php:
```diff
@@ -27,19 +27,24 @@
  * Class generates "By order status" form
  * in "Sell > Orders > Invoices" page.
+ *
+ * Backwards compatibility break introduced in 1.7.8.0 due to extension of TranslationAwareType
```
```suggestion
 * Backwards compatibility break introduced in 8.0.0 due to extension of TranslationAwareType
```

matks commented on templates/bundles/PrestaShopBundle/Admin/Sell/Order/Invoices/Blocks/invoice_options.html.twig:
```diff
@@ -35,97 +35,9 @@
-            </div>
-            {{ form_rest(invoiceOptionsForm) }}
+            {% block invoice_options_form_widget %}
+              {{ form_widget(invoiceOptionsForm) }}
```
I always love this part

matks commented on tests/Unit/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsDataProviderTest.php:
```diff
@@ -0,0 +1,156 @@
+        ];
+        $exceptionThrown = false;
+        $configurationError = new InvoiceConfigurationError(InvoiceConfigurationError::ERROR_INCORRECT_INVOICE_NUMBER, InvoiceOptionsType::FIELD_INVOICE_NUMBER);
+        try {
```
PHPUnit provides some assertions for exceptions. Did you choose not to use them on purpose?

matks replied:
https://phpunit.readthedocs.io/en/latest/writing-tests-for-phpunit.html#writing-tests-for-phpunit-exceptions-examples-exceptiontest-php

JevgenijVisockij replied:
I will read the doc, but the reason I chose not to use it is that I want to check my exception for correct configuration errors. I am not sure if PHP unit supports that functionality.

matks reviewed (commented):
Hi @JevgenijVisockij I reviewed the code.

Everything looks good (I provided some nice-to-have comments but nothing blocking). However I can see you introduce some new concepts: ConfigurationErrors, ErrorCollections, ErrorFactories... could you please explain the PR body what responsibility each class holds?

For example if tomorrow we want to apply the same concept to "Webservice" configuration form, what do we need to add? One ErrorFactory? One WebserviceDataError ?

I think the context could also inside the interfaces :) what goal it serves, what is expected input, where is it supposed to be used...

JevgenijVisockij committed:
Apply suggestions from code review

Changing deprecation notice to correct version.

Co-authored-by: Mathieu Ferment <mathieu.ferment@prestashop.com>

Progi1984 commented on src/PrestaShopBundle/Form/ErrorMessage/Factory/AdministrationConfigurationErrorFactory.php:
```diff
@@ -0,0 +1,86 @@
+     *
+     * @return string
+     */
+    public function getErrorMessageForConfigurationError(ConfigurationErrorInterface $error, string $label): ?string
```
Could you add unit test for checking the behavior ? Thanks

JevgenijVisockij replied:
Added the test

Progi1984 commented on src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoicesByDateDataProvider.php:
```diff
@@ -65,47 +68,50 @@ public function getData()
+     *
+     * @return void
      */
     private function validate(array $data)
```
Could you add unit test for checking behavior ?

JevgenijVisockij replied:
Added the test

Progi1984 reviewed (commented):
@JevgenijVisockij Thank you for your contribution. I added two last comments (unit tests and it's ok for me).

matks commented:
Code is OK for me, missing the 2 tests requested by @Progi1984 😉 

Also could it be possible Everything looks good to explain in the PR body what responsibility each new class holds?

See https://github.com/PrestaShop/PrestaShop/pull/20243#pullrequestreview-731963982

JevgenijVisockij committed:
Added integration tests for date provider and administration configuration error factory

JevgenijVisockij committed:
ADded php unit test for validator

# Conflicts:
#	src/Core/ConstraintValidator/NoTagsValidator.php
#	tests/Unit/Core/ConstraintValidator/NoTagsValidatorTest.php

# Conflicts:
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form_data_provider.yml

JevgenijVisockij committed:
Code review fixes

# Conflicts:
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form_data_provider.yml

JevgenijVisockij committed:
Removed changes from invoices controller

# Conflicts:
#	src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php

Progi1984 commented on tests/Unit/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceByDateDataProviderTest.php:
```diff
@@ -0,0 +1,102 @@
+        $orderInvoiceByDateDataProvider->setData($data);
+    }
+
+    public function getInvoiceNoHtmlFields(): array
+    {
+        return [
+            [InvoiceOptionsType::FIELD_INVOICE_PREFIX],
+            [InvoiceOptionsType::FIELD_LEGAL_FREE_TEXT],
+            [InvoiceOptionsType::FIELD_FOOTER_TEXT],
+        ];
+    }
```
```suggestion
```

Progi1984 reviewed (commented):
@JevgenijVisockij A small feedback and it's good

sowbiba commented on src/Core/Form/ErrorMessage/Factory/ConfigurationErrorFactoryInterface.php:
```diff
@@ -0,0 +1,40 @@
+     * @param ConfigurationErrorInterface $error
+     * @param string $label
+     *
+     * @return string
```
```suggestion
     * @return string|null
```

sowbiba commented on src/PrestaShopBundle/Form/ErrorMessage/Factory/AdministrationConfigurationErrorFactory.php:
```diff
@@ -0,0 +1,86 @@
+     * @param ConfigurationErrorInterface $error
+     * @param string $label
+     *
+     * @return string
```
```suggestion
     * @return string|null
```

sowbiba commented on src/PrestaShopBundle/Form/ErrorMessage/Factory/ConfigurationErrorFactory.php:
```diff
@@ -0,0 +1,108 @@
+            }
+        }
+
+        return $messages;
```
you'll only have the first error message thrown. Is it assumed ?

sowbiba commented on src/PrestaShopBundle/Form/ErrorMessage/Factory/InvoiceConfigurationErrorFactory.php:
```diff
@@ -0,0 +1,103 @@
+     * @param ConfigurationErrorInterface $error
+     * @param string $label
+     *
+     * @return string
```
```suggestion
     * @return string|null
```

sowbiba commented on src/PrestaShopBundle/Form/ErrorMessage/LabelProvider.php:
```diff
@@ -0,0 +1,69 @@
+        foreach ($view->children as $child) {
+            if ($fieldName === $child->vars['name']) {
+                if (!isset($child->vars['label'])) {
+                    throw new FieldNotFoundException(
```
The field is found, it's the field label which is not set. The exception name can lead to confusion

sowbiba reviewed (commented):
some small feedbacks. Good job @JevgenijVisockij

sowbiba committed:
Rebase, cs-fixer and YAML lint fixes

matks reviewed (commented):
Nothing to add to @sowbiba review 👍

Progi1984 reviewed (commented):
A last feedback in `tests/Unit/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceByDateDataProviderTest.php`

sowbiba committed:
Specify exception type, fix typehinting

Progi1984 reviewed (commented):
@JevgenijVisockij Thank you for your contribution. Small feedbacks from me & @sowbiba

JevgenijVisockij committed:
Update src/PrestaShopBundle/Form/ErrorMessage/LabelProvider.php

Added extra validation for fieldname

Co-authored-by: Progi1984 <progi1984@gmail.com>

JevgenijVisockij committed:
Fixed label provider test

khouloudbelguith commented:
Hi @JevgenijVisockij,

I have some issues:

**First issue**
The invoices page is not well displayed
![image](https://user-images.githubusercontent.com/16067358/149347088-80b5bbeb-1f6d-4549-a2c0-d8af596502a1.png)

**It should be**
![image](https://user-images.githubusercontent.com/16067358/149347308-ba609cb3-3541-4eab-9b5c-ed9c073e4326.png)


**Second issue**
The field `Invoice prefix`, we can set this text with a too long-expression,
1. for example set this text `tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt` and save
2. Go to BO > Orders > TRy to download an invoice
3. The invoice cannot be downloaded
4. View an order
5. In the document tab > See the issue 
![image](https://user-images.githubusercontent.com/16067358/149348187-e03d0b5d-271b-4c87-8aca-fce2af56ef1d.png)

**Third issue**
1. Go to BO > Orders > Invoices page
2. Set the invoice number with invalid characters for example `>>>>`
3. Save
4. See error: a green alert is displayed, we should have a red alert.

https://user-images.githubusercontent.com/16067358/149349317-d5689446-baa9-4619-a2e7-89441b105156.mp4

Thanks!

Progi1984 commented:
Hi @JevgenijVisockij,
Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.
Thanks!

Progi1984 commented:
@jolelievre I take it.

matks commented:
Hello all,

I'm really sorry but following the [8.0.0 freeze](https://build.prestashop.com/news/prestashop-feature-freeze-v8/) happened and now until further notice `develop` branch is going to become the branch for 8.1, which means we cannot accept BC breaks like this one 😞 I'm sorry

I see 2 solutions:
1. Wait for PrestaShop 9 to refactor this form
2. Create two forms and twig templates, and use a Feature Flag to trigger the use of the refactored form

JevgenijVisockij committed:
Simplified the invoice controller forms

# Conflicts:
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form_type.yml

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php
#	src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Invoices/Blocks/generate_by_date.html.twig
#	src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Invoices/Blocks/generate_by_status.html.twig
#	src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Invoices/Blocks/invoice_options.html.twig

JevgenijVisockij committed:
CS fixer fixes

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
Added better validation to Invoices Options form

# Conflicts:
#	src/Core/ConstraintValidator/Constraints/NoTags.php
#	src/Core/ConstraintValidator/NoTagsValidator.php

# Conflicts:
#	src/PrestaShopBundle/Controller/Admin/Sell/Order/InvoicesController.php
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form_data_provider.yml

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
CS fixer fixes

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
Fixed some comments

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
Removed uneeded required => true

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
Fixed validation for invoices

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig

# Conflicts:
#	templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php
#	src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig

JevgenijVisockij committed:
Added wording changes regarding invoice number

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
Removed uneeded param in administration controller

# Conflicts:
#	src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php

JevgenijVisockij committed:
Added errors to invoiceOptionsDataProvider

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
Added language to errors

# Conflicts:
#	src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php

JevgenijVisockij committed:
Removed no tags constraints since they are no longer needed here

# Conflicts:
#	src/Core/ConstraintValidator/NoTagsValidator.php

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
Moved error message generation to separate service

# Conflicts:
#	src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form.yml

# Conflicts:
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form.yml

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
Changed invoice type from number to integer

# Conflicts:
#	src/PrestaShopBundle/Form/Admin/Sell/Order/Invoices/InvoiceOptionsType.php

JevgenijVisockij committed:
Removed no longer needed error message retrieval

# Conflicts:
#	src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/AdministrationController.php

JevgenijVisockij committed:
Removed unneeded number widget, Moved InvalidConfigurationErrorMessageFactory.php

# Conflicts:
#	templates/bundles/PrestaShopBundle/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig

# Conflicts:
#	src/PrestaShopBundle/Resources/config/services/bundle/form/form.yml
#	src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig

JevgenijVisockij committed:
Apply suggestions from code review

Updated twig routes

Co-authored-by: atomiix <tbaccelli@gmail.com>
# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/Sell/Order/Invoices/Blocks/generate_by_date.html.twig

JevgenijVisockij committed:
Made changes to ConfigurationErrorFactory to separate it into multiple classes and make it extendable by modules

# Conflicts:
#	src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig

JevgenijVisockij committed:
Fixes after rebase

JevgenijVisockij committed:
Adding back NoTagsValidatorTest.php

JevgenijVisockij committed:
Fixing to the correct test

JevgenijVisockij committed:
Removed uneeded constraint

JevgenijVisockij commented on src/PrestaShopBundle/Form/Exception/DataProviderException.php:
```diff
@@ -29,26 +29,27 @@
      */
-    public function getInvalidConfigurationDataErrors(): InvalidConfigurationDataErrorCollection
+    public function getConfigurationErrors(): ConfigurationErrorCollection
     {
```
I guess this also considered BC break, should I change it back for the sake of simplicity?

JevgenijVisockij committed:
CS fixer fixes

JevgenijVisockij commented:
> Hi @JevgenijVisockij,
> 
> I have some issues:
> 
> **First issue** The invoices page is not well displayed ![image](https://user-images.githubusercontent.com/16067358/149347088-80b5bbeb-1f6d-4549-a2c0-d8af596502a1.png)
> 
> **It should be** ![image](https://user-images.githubusercontent.com/16067358/149347308-ba609cb3-3541-4eab-9b5c-ed9c073e4326.png)
> 
> **Second issue** The field `Invoice prefix`, we can set this text with a too long-expression,
> 
>     1. for example set this text `tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt` and save
> 
>     2. Go to BO > Orders > TRy to download an invoice
> 
>     3. The invoice cannot be downloaded
> 
>     4. View an order
> 
>     5. In the document tab > See the issue
>        ![image](https://user-images.githubusercontent.com/16067358/149348187-e03d0b5d-271b-4c87-8aca-fce2af56ef1d.png)
> 
> 
> **Third issue**
> 
>     1. Go to BO > Orders > Invoices page
> 
>     2. Set the invoice number with invalid characters for example `>>>>`
> 
>     3. Save
> 
>     4. See error: a green alert is displayed, we should have a red alert.
> 
> 
> Invoices.PrestaShop.mp4
> 
> Thanks!

Regarding bugs 1 is fixed, 
2 Technically it just doesn't look great when prefix is too large, but that's on user to not set a huge prefix. I can set a limit but what should it be?
3. @jolelievre I am not sure if I can fix it. Because type is integer Symfony converts any text to 0 by the time validation happens, so I can't really validate it.

> Hi @JevgenijVisockij,
> 
> I have some issues:
> 
> **First issue** The invoices page is not well displayed ![image](https://user-images.githubusercontent.com/16067358/149347088-80b5bbeb-1f6d-4549-a2c0-d8af596502a1.png)
> 
> **It should be** ![image](https://user-images.githubusercontent.com/16067358/149347308-ba609cb3-3541-4eab-9b5c-ed9c073e4326.png)
> 
> **Second issue** The field `Invoice prefix`, we can set this text with a too long-expression,
> 
>     1. for example set this text `tttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt` and save
> 
>     2. Go to BO > Orders > TRy to download an invoice
> 
>     3. The invoice cannot be downloaded
> 
>     4. View an order
> 
>     5. In the document tab > See the issue
>        ![image](https://user-images.githubusercontent.com/16067358/149348187-e03d0b5d-271b-4c87-8aca-fce2af56ef1d.png)
> 
> 
> **Third issue**
> 
>     1. Go to BO > Orders > Invoices page
> 
>     2. Set the invoice number with invalid characters for example `>>>>`
> 
>     3. Save
> 
>     4. See error: a green alert is displayed, we should have a red alert.
> 
> 
> Invoices.PrestaShop.mp4
> 
> Thanks!

1 and 3 should be fine now, regarding second I am not sure if that's a problem. It's on user to not set huge invoice prefix, but it can be added no problem any idea what the limit should be?

JevgenijVisockij committed:
Removed uneeded code from AdministrationController.php

JevgenijVisockij commented:
> Hello all,
> 
> I'm really sorry but following the [8.0.0 freeze](https://build.prestashop.com/news/prestashop-feature-freeze-v8/) happened and now until further notice `develop` branch is going to become the branch for 8.1, which means we cannot accept BC breaks like this one disappointed I'm sorry
> 
> I see 2 solutions:
> 
>     1. Wait for PrestaShop 9 to refactor this form
> 
>     2. Create two forms and twig templates, and use a Feature Flag to trigger the use of the refactored form
Well for type change part, only construct is changed, according to ADR that's okay right? 
https://github.com/PrestaShop/ADR/blob/master/0017-backward-compatibility-promise.md
I think that leaves only one BC break I wrote comment about earlier(https://github.com/PrestaShop/PrestaShop/pull/20243#discussion_r953709160), if it too cannot pass, changing it back wouldn't be a problem.

JevgenijVisockij committed:
PHP Static Analytics fix

JevgenijVisockij committed:
PHP Static Analytics fixes

JevgenijVisockij committed:
Changing error name to avoid bc break

JevgenijVisockij committed:
Changing error name to avoid bc break

JevgenijVisockij commented:
This is blocked by https://github.com/PrestaShop/PrestaShop/pull/29494 since it relies on the same option error system that is improved in the PR
https://github.com/PrestaShop/PrestaShop/pull/29494
